kernel: atomics: support for 64-bit atomic operations

cfriedt opened the pull request:
This change adds support for 64-bit atomic operations for any machine where `sizeof(long) == 8`.

Fixes #39530
Depends-on
* thesofproject/sof#4975
* zephyrproject-rtos/sof#14

andyross reviewed (commented):
Not sure how I feel about making atomic_t a variable-sized type.  Seems like it would be safer and more portable[1] (but obviously a lot more boilerplate) to have two types and two sets of functions?

Also: a wart you might want to fix while you're in here is that atomic_t has always been a typedef and not an opaque struct.  There have been **many** situations noticed over the years where code was incorrectly modifying them in an unsynchronized way (which then begets discussion in review about whether or not the locking environment makes that safe or not, etc...).

[1] Consider the case of a 64 bit app that works fine being built on a 32 bit platform and then failing due to synchronization overflows or signedness bugs that were invisible on the larger word.

andyross commented:
I'm the wrong person to talk to about language standards.  But certainly the release after an LTS is the perfect time to revisit API changes.  I wouldn't cry to see the existing layer be junked (though we'd absolutely want to keep the arch layer abstractions available, because I don't trust toolchains any farther than I can throw them).

cfriedt commented:
> I'm the wrong person to talk to about language standards. But certainly the release after an LTS is the perfect time to revisit API changes. I wouldn't cry to see the existing layer be junked (though we'd absolutely want to keep the arch layer abstractions available, because I don't trust toolchains any farther than I can throw them).

https://github.com/zephyrproject-rtos/zephyr/issues/39641

cfriedt commented:
More print formatting errors - will fix shortly.

cfriedt committed:
kernel: atomics: support for 64-bit atomic operations

With this change, `atomic_t` is 32-bit for 32-bit architectures
and 64-bit for 64-bit architectures. More specifically,
`sizeof(atomic_t) == sizeof(long)`.

Fixes #39530

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
doc kernel: atomics: support for 64-bit atomic operations

By redefining `atomic_t` as `long`, the type is 32-bit on
32-bit architectures and 64-bit on 64-bit architectures.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
kernel: atomics: update print specifiers for atomic_t

The print specifier for `atomic_t` should be updated
to `%ld`, `%lu`, or `%lx` to account for the type
change of `atomic_t` to `long`.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
drivers: serial: nrfx: refactor for atomic_t as long

This driver aliases a regular `int` to `atomic_t` but that
should be updated to `long` with the change to `atomic_t`.

Added a comment to highlight that the variable was aliased.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
portability: cmsis: use correct type for atomics

With the change of `atomic_t` from `int` to `long` there
were a few places where there was some type aliasing
occuring.

Update CMSIS to use `atomic_t` rather than `int` for all
atomic operations.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
west.yml: sof: update audio/sof for long atomic_t

The audio/sof Zephyr module needed print specifiers updated
for the transition from `int` to `long` for `atomic_t`
which supports 32-bit atomics and also 64-bit atomics.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: kernel: atomics: expand atomic operations to 64-bit

This change updates the atomic tests to validate 32-bits on
32-bit architectures and 64-bits on 64-bit architectures.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

lgirdwood commented:
@nashif @cfriedt SOF PR now merged.

cfriedt commented:
update sof hash after thesofproject/sof#4975 was merged

cfriedt commented:
rebased

cfriedt committed:
Bluetooth: Audio: otc: update print format specifier for atomics

Atomics changed from `int` to `long` and print format specifiers
must be updated as well.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
updated `subsys/bluetooth/audio/otc.c` to use "%lu"
